tests/test_databases.py

```diff
@@ -477,6 +480,212 @@ async def test_transaction_commit(database_url):
+    del database
+    gc.collect()
+
+    # Should have dropped reference to connection, even without proper cleanup
+    open_connections = _ACTIVE_CONNECTIONS.get()
+    assert isinstance(open_connections, MutableMapping)
+    assert len(open_connections) == 0
+
+
+@pytest.mark.parametrize("database_url", DATABASE_URLS)
+@async_adapter
+async def test_transaction_context_cleanup_contextmanager(database_url):
+    """
+    Ensure that contextvar transactions are not persisted unecessarily.
+    """
+    from databases.core import _ACTIVE_TRANSACTIONS
+
+    assert _ACTIVE_TRANSACTIONS.get() is None
+
+    async with Database(database_url) as database:
+        async with database.transaction() as transaction:
+
+            open_transactions = _ACTIVE_TRANSACTIONS.get()
+            assert isinstance(open_transactions, MutableMapping)
+            assert open_transactions.get(transaction) is transaction._transaction
+
+        # Context manager closes, open_transactions is cleaned up
+        open_transactions = _ACTIVE_TRANSACTIONS.get()
+        assert isinstance(open_transactions, MutableMapping)
+        assert open_transactions.get(transaction, None) is None
+
+
+@pytest.mark.parametrize("database_url", DATABASE_URLS)
+@async_adapter
+async def test_transaction_context_cleanup_garbagecollector(database_url):
+    """
+    Ensure that contextvar transactions are not persisted unecessarily, even
+    if exit handlers are not called.
+
+    This test should be an XFAIL, but cannot be due to the way that is hangs
+    during teardown.
+    """
+    from databases.core import _ACTIVE_TRANSACTIONS
+
+    assert _ACTIVE_TRANSACTIONS.get() is None
+
+    async with Database(database_url) as database:
+        transaction = database.transaction()
+        await transaction.start()
+
+        # Should be tracking the transaction
+        open_transactions = _ACTIVE_TRANSACTIONS.get()
+        assert isinstance(open_transactions, MutableMapping)
+        assert open_transactions.get(transaction) is transaction._transaction
+
+        # neither .commit, .rollback, nor .__aexit__ are called
+        del transaction
+        gc.collect()
+
+        # TODO(zevisert,review): Could skip instead of using the logic below
+        # A strong reference to the transaction is kept alive by the connection's
+        # ._transaction_stack, so it is still be tracked at this point.
+        assert len(open_transactions) == 1
+
+        # If that were magically cleared, the transaction would be cleaned up,
+        # but as it stands this always causes a hang during teardown at
+        # `Database(...).disconnect()` if the transaction is not closed.
+        transaction = database.connection()._transaction_stack[-1]
+        await transaction.rollback()
+        del transaction
+
+        # Now with the transaction rolled-back, it should be cleaned up.
+        assert len(open_transactions) == 0
```

Comment:
I got a bit blocked with this test, as I noted in these comments. I am trying to test that _if a transaction is aborted without calling any teardown logic_ then the weak map in the context variable does not keep a reference to it.

`databases.core.Connection` objects have a stack (using `list`) of transactions that have been opened from them, and that stack is managed through exit handing code. If we're not calling that in this test, then there will still be a strong reference to the transaction and it won't get garbage collected. 

On one hand, I think this is programming error and it's not on _**databases**_ to account for. On the other, you could make the same argument about the use of `WeakKeyDictionary` to avoid blocking the garbage collector from cleaning up these transactions. All database drivers will hang indefinitely during test teardown `transaction.rollback()` was not called. 

I can think of two approaches here:
- Add a finalizer to `databases.core.Transaction` to call `.rollback` if nothing else has by the time they are being garbage collected. This probably requires the next approach as well though for objects to actually have no references and get gc'd..
- We could maybe use a `WeakValueDictonary` like a list (keys of incrementing integers) for the `_transaction_stack`, so that if a transaction is somehow dropped by a user, _**databases**_ can still clean up. This feels a bit "hairy" though, and I'd be more inclined to call this user-error..

Reply:
Hm I'm not sure I entirely follow here, but it sounds bad to hang on shutdown. Is that a change in behavior?

Reply:
This is not new behaviour, I've triple-checked that. This is existing behaviour that has been exposed through these new tests that are checking the cleanup functionality of weakref-based collections. It is not introduced, changed, or fixed in this PR - the effect here is that I cannot completely test that references to `databases.core.Transaction` instances are discarded by variables going out of scope.

While this of course isn't on the happy path - scenarios where a transaction instance is deleted without calling `.commit()` or `.rollback()` are possible if users are using the [documented low-level transaction management API](https://www.encode.io/databases/connections_and_transactions/#:~:text=For%20a%20lower%2Dlevel%20transaction%20API%3A), and they didn't set up their try/except/else logic correctly (this is why I consider this user-error).

```py
# MVE that hangs on master, 0.7.0, 0.6.2, 0.5.5, 0.4.3, 0.3.2, 0.2.6, 0.1.12 
import databases, asyncio, os

async def mve():
    async with databases.Database(os.environ['TEST_DATABASE_URL']) as db:
        transaction = db.transaction()
        await transaction.start()

asyncio.run(mve())
```

It also makes sense that this hangs on shutdown. From the database driver's perspective, there is an open transaction that has not executed a `COMMIT` or `ROLLBACK` statement, better wait for that to happen before shutting down.

Reply:
👍 as long as it's not a regression I think it makes sense to track it in a new issue to determine the best solution. I think that ideally we'd warn that a transaction was left open rather than hanging forever.

Reply:
It might not even need to be a new issue. I can try to see if there's something calling this out already, otherwise I'll create a new issue.

Reply:
Follow up issue: #554